image_view/src/image_saver_node.cpp

```diff
@@ -0,0 +1,258 @@
+  //  3. flag 'save_all_image'.
+  if (!save_image_service && request_start_end) {
+    if (start_time_ == rclcpp::Time(0)) {
+      return;
+    } else if (start_time_ > image_msg->header.stamp) {
+      return;  // wait for message which comes after start_time
+    } else if ((end_time_ != rclcpp::Time(0)) && (end_time_ < image_msg->header.stamp)) {
+      return;  // skip message which comes after end_time
+    }
+  }
+
+  // save the image
+  std::string filename;
+  if (!saveImage(image_msg, filename)) {
+    return;
+  }
+
+  count_++;
+}
+
+void ImageSaverNode::callbackWithCameraInfo(
+  const sensor_msgs::msg::Image::ConstSharedPtr & image_msg,
+  const sensor_msgs::msg::CameraInfo::ConstSharedPtr & info)
+{
+  has_camera_info_ = true;
+
+  if (!save_image_service && request_start_end) {
+    if (start_time_ == rclcpp::Time(0)) {
+      return;
+    } else if (start_time_ > image_msg->header.stamp) {
+      return;  // wait for message which comes after start_time
+    } else if ((end_time_ != rclcpp::Time(0)) && (end_time_ < image_msg->header.stamp)) {
+      return;  // skip message which comes after end_time
+    }
+  }
+
+  // save the image
+  std::string filename;
+  if (!saveImage(image_msg, filename)) {
+    return;
+  }
+
+  // save the CameraInfo
+  if (info) {
+    filename = filename.replace(filename.rfind("."), filename.length(), ".ini");
+    camera_calibration_parsers::writeCalibration(filename, "camera", *info);
+  }
+
+  count_++;
+}
+
+ImageSaverNode::ImageSaverNode(const rclcpp::NodeOptions & options)
+  : rclcpp::Node("image_saver_node", options),
+    is_first_image_(true), has_camera_info_(false), count_(0)
+{
+  auto topic = rclcpp::expand_topic_or_service_name("image", this->get_name(), this->get_namespace());
+
+  // Useful when CameraInfo is being published
+  cam_sub_ = image_transport::create_camera_subscription(
+    this, topic, std::bind(
+      &ImageSaverNode::callbackWithCameraInfo, this, std::placeholders::_1, std::placeholders::_2),
+    "raw");
+
+  // Useful when CameraInfo is not being published
+  image_sub_ = image_transport::create_subscription(
+    this, topic, std::bind(
+      &ImageSaverNode::callbackWithoutCameraInfo, this, std::placeholders::_1),
+    "raw");
+
+  std::string format_string;
+  format_string = this->declare_parameter("filename_format", std::string("left%04i.%s"));
+  encoding = this->declare_parameter("encoding", std::string("bgr8"));
+  save_all_image = this->declare_parameter("save_all_image", true);
+  request_start_end = this->declare_parameter("request_start_end", false);
+  g_format.parse(format_string);
+
+  if (request_start_end && !save_all_image) {
+    RCLCPP_WARN(
+      this->get_logger(), "'request_start_end' is true, so overwriting 'save_all_image' as true");
+  }
+
+  save_srv_ = this->create_service<std_srvs::srv::Empty>(
+    "save", std::bind(&ImageSaverNode::service, this, _1, _2, _3));
+
+  // FIXME(unkown): This does not make services appear
```

Comment:
Is this still a fix need?

Reply:
Instead of trying to "fix" this, I just removed the `request_start_end` parameter. Having topics or services appear or not appear based on parameters is bad form. I just have both services available all the time and people can choose how to use the node (whether they use the `save` service or the `start` and `end` services).